### PR TITLE
feat: add braces assist

### PR DESCRIPTION
**Pull request by vasilev-alex**
This assist adds support for adding braces where it may be appropriate (e.g. lambda expressions)

ex

### CODE REVIEW

File: crates/ide-assists/src/handlers/add_braces.rs [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+//     }
+// }
+// ```
+pub(crate) fn add_braces(acc: &mut Assists, ctx: &AssistContext<'_>) -> Option<()> {
+    let (expr_type, expr) = get_replacement_node(ctx)?;
+
+    acc.add(
+        AssistId("add_braces", AssistKind::RefactorRewrite),
+        match expr_type {
+            ParentType::ClosureExpr => "Add braces to lambda expression",
```
Comment (lowr) [inline crates/ide-assists/src/handlers/add_braces.rs:37]:
```suggestion
            ParentType::ClosureExpr => "Add braces to closure body",
```

I think Rust prefers the term "closure" over "lambda"?

File: crates/ide-assists/src/handlers/add_braces.rs [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+}
+
+enum ParentType {
+    MatchArmExpr,
+    ClosureExpr,
+}
+
+fn get_replacement_node(ctx: &AssistContext<'_>) -> Option<(ParentType, ast::Expr)> {
+    if let Some(match_arm) = ctx.find_node_at_offset::<ast::MatchArm>() {
+        let match_arm_expr = match_arm.syntax().children().find_map(ast::Expr::cast)?;
```
Comment (lowr) [inline crates/ide-assists/src/handlers/add_braces.rs:61]:
```suggestion
        let match_arm_expr = match_arm.expr()?;
```

File: crates/ide-assists/src/handlers/add_braces.rs [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+            let block_expr = AstNodeEdit::indent(
+                &make::block_expr(None, Some(expr.clone())),
+                AstNodeEdit::indent_level(&expr),
+            );
+
+            builder.replace(expr.syntax().text_range(), block_expr.syntax().text());
+        },
+    );
+
+    Some(())
```
Comment (lowr) [inline crates/ide-assists/src/handlers/add_braces.rs:51]:
```suggestion
    )
```

**Review by lowr — COMMENTED** [review]
Leaving some cosmetic suggestions. (not sure if this is worth having, but I feel it comes in handy at times)

Also, I think the expressions we're wrapping in braces need to be escaped. Can you escape using [`utils::escape_non_snippet()`](https://github.com/rust-lang/rust-analyzer/blob/9a6294d7038e7eab00beafdf64ec4aa50a4c66a2/crates/ide-assists/src/utils.rs#L216) and add a test like [this one](https://github.com/rust-lang/rust-analyzer/blob/9a6294d7038e7eab00beafdf64ec4aa50a4c66a2/crates/ide-assists/src/handlers/move_const_to_impl.rs#L448)?

File: crates/ide-assists/src/handlers/add_braces.rs (hunk 1 above)
Comment (vasilev-alex) [reply crates/ide-assists/src/handlers/add_braces.rs:37]: yeah, I wasn't sure about the name as well

**Comment by vasilev-alex** [comment]
@lowr not sure if I understand the suggestion about escaping. Could you provide some examples of why it's necessary in the case of wrapping expressions?

**Comment by lowr** [comment]
Apologies, we need to escape stuffs only if we're using snippets and apparently this is not the case! I was misunderstanding about escaping, I'm starting to double if `utils::escape_non_snippet()` is of any use at all then...

**Comment by vasilev-alex** [comment]
@Veykril do you think this can be merged?

**Review by Veykril — APPROVED** [review]

**Comment by Veykril** [comment]
I am also not 100% whether this is really worth to have (there are other ways to em-brace expressions like this, selection + typing `{` in most editors), but this is self-contained so we can always remove it later if it is deemed to not be used much.
Thanks!
@bors r+

**Pull request merged**